src/main/scala/loamstream/conf/Locations.scala

```diff
@@ -1,31 +1,57 @@
 package loamstream.conf
 
-import java.nio.file.Paths
 import java.nio.file.Path
 
 /**
  * @author clint
  * Dec 17, 2018
- * 
+ *
  * An object containing the location of various things: output files, default log dir, etc.
  */
+trait Locations {
+  def loamstreamDir: Path
+
+  def jobDataDir: Path
+  
+  def logDir: Path
+
+  def dryRunOutputFile: Path
+
+  def dbDir: Path
+
+  def ugerDir: Path
+  def lsfDir: Path
+}
+
 object Locations {
   
   import java.nio.file.Files.createDirectories
+  import java.nio.file.Paths.{ get => path }
+  import loamstream.util.Paths.Implicits._
   
-  val loamstreamDir: Path = createDirectories(Paths.get("./.loamstream"))
+  final case class Literal(
+    loamstreamDir: Path = Default.loamstreamDir,
+    jobDataDir: Path = Default.jobDataDir,
+    logDir: Path = Default.logDir,
+    dryRunOutputFile: Path = Default.dryRunOutputFile,
+    dbDir: Path = Default.dbDir,
+    ugerDir: Path = Default.ugerDir,
+    lsfDir: Path = Default.lsfDir) extends Locations
   
-  val logDir: Path = createDirectories(loamstreamDir.resolve("logs").normalize)
-    
-  val dryRunOutputFile: Path = logDir.resolve("joblist").normalize
+  object Default extends Locations {
+    override val loamstreamDir: Path = createDirectories(path("./.loamstream"))
+
+    private[this] val jobDir: Path = createDirectories((loamstreamDir / "jobs").normalize)
     
-  val jobOutputDir: Path = createDirectories(loamstreamDir.resolve("job-outputs").normalize)
-  
-  val dbDir: Path = createDirectories(loamstreamDir.resolve("db"))
-  
-  lazy val ugerDir = createDirectories(loamstreamDir.resolve("uger").normalize)
-  lazy val lsfDir = createDirectories(loamstreamDir.resolve("lsf").normalize)
-  
-  lazy val ugerScriptDir: Path = createDirectories(ugerDir.resolve("scripts").normalize)
-  lazy val lsfScriptDir: Path = createDirectories(lsfDir.resolve("scripts").normalize)
+    override val jobDataDir: Path = createDirectories(jobDir / "data")
+
+    override val logDir: Path = createDirectories((loamstreamDir / "logs").normalize)
```

Comment:
Just curious if `createDirectories` actually creates the directories (on disk) or just builds the path name up? If it actually makes them on disk, then I'd vote for something other than this being a `val` and having directories get magically made just by an Object being loaded into memory.

If it doesn't make them, then would another name be warranted (`joinPathName` or something)?

Reply:
> Just curious if createDirectories actually creates the directories (on disk) or just builds the path name up? If it actually makes them on disk, then I'd vote for something other than this being a val and having directories get magically made just by an Object being loaded into memory. To me that's a definite side-effect that should explicitly done by the programmer.

Yeah, this is nothing but a bunch of nasty side effects.  (`createDirectories` creates directories.)  I don't like it, and at one point tried out `Locations` being a bunch of paths describing places on the FS, and a side-effecting method somewhere else that created all the paths in a `Locations` object, presumably at the start of a run.

I remember one of the catches being that this made testing a pain, but I can't remember why exactly.  A long time ago I made it so most tests that write anything to the FS do some in temp dirs with guaranteed-unique names that are automatically cleaned up, so a lot (all?) of the needed parameterization is in place.  

Looking at this code more, it definitely needs to get revised.  There are a couple of small/medium gotchas in `Locations`, which suggests more would show up after a closer look.  I'd like to tackle this in another PR though, since the needed refactorings would be moderately-sized, and this PR is already huge.

Reply:
Good news: I was able to get rid of the calls to `createDirectories` in `Locations`, so that class is no longer side-effecting.